[serve.llm] Refactor/Consolidate LoRA downloading

nrghosh opened the pull request:
Move LoraModelLoader to download_utils and enhance download_lora_adapter to unify LoRA downloading functionality. Eliminates some duplication while keeping backward compatibility.

- Centralize download utilities in common/utils vs serve-specific code
- Unify retry logic, caching, and file locking across LoRA downloads
- Enhanced download_lora_adapter supports both simple and advanced modes
- Simple mode: Basic download using original CloudModelDownloader (backward compatible)
- Advanced mode: Uses LoraModelLoader with disk caching, retry logic, and configurable timeouts
- Update tests
- Fix linting
- Resolves [TODO](https://github.com/ray-project/ray/blob/62e3169e3d2709df03d16606a806f9d28cd709e6/python/ray/llm/_internal/common/utils/download_utils.py#L301)


<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?




<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

nrghosh committed:
refactor(serve.llm): Consolidate LoRA downloading

Move LoraModelLoader to download_utils and enhance download_lora_adapter
to unify LoRA downloading functionality. Eliminates code duplication while
maintaining backward compatibility.

- Centralize download utilities in common/utils vs serve-specific code
- Unify retry logic, caching, and file locking across LoRA downloads
- Enhanced download_lora_adapter supports both simple and advanced modes
- Simple mode: Basic download using original CloudModelDownloader (backward compatible)
- Advanced mode: Uses LoraModelLoader with disk caching, retry logic, and configurable timeouts
- Update tests
- Fix linting

Resolves TODO to unify lora_model_loader functionality.

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

eicherseiji commented:
Serve tests are inclusive due to https://github.com/anyscale/product/pull/34303/

nrghosh commented:
> Serve test [failures](https://buildkite.com/ray-project/release/builds/45233) are inclusive due to [anyscale/product#34303](https://github.com/anyscale/product/pull/34303). More context in [Slack thread](https://anyscaleteam.slack.com/archives/C02J07KKF2A/p1749676258456189).

Should have been fixed by [PR](https://github.com/anyscale/product/pull/34303/files) - will re-trigger

kouroshHakha commented on python/ray/llm/_internal/common/utils/download_utils.py:
```diff
@@ -316,7 +326,329 @@ def download_lora_adapter(
+            raise ValueError(f"max_tries must be >=1, got {max_tries}")
+        self.max_tries = max_tries
+
+    async def load_model(
```
so LLMServer needs to use this guy now, Where is that change in this PR?

nrghosh replied:
- updating LLMServer to use the new private _LoraModelLoader --> injecting the loader through constructor

nrghosh committed:
Merge branch 'master' into refactor-lora_model_loader

Signed-off-by: Nikhil G <nrghosh@users.noreply.github.com>

kouroshHakha commented on python/ray/llm/_internal/common/utils/download_utils.py:
```diff
@@ -10,9 +13,13 @@
     is_remote_path,
 )
-from ray.llm._internal.common.utils.import_utils import try_import
+
```
having `DiskMultiplexConfig` in `server_models` does not make sense anymore. It should become part of this common/utils.py package and other places that used it should import it. It's a private data structure that makes sense in the context of LoRAModelDownloader.

nrghosh replied:
addressed:
- moving DiskMultiplexConfig to common/models.py
- making it a shared data structure between serve and batch
- getting rid of serve-specific dependencies

kouroshHakha reviewed (changes requested):
This is not the desired level of refactoing. The multiplex/util.py stuff are still there. We need to do a comprehensive refactor to truly clean up and unify the api requirements between batch and serve. Backward compatibility is not of concern here since we are dealing with an internal refactor.


So think of it this way. You need to go generic -> specific. First think of what are the generic primitives that I need to keep and expand their functionality in a general way so that I can build the other things on top.

- `common/download_utils.py` --> Should include abstraction that make it easy to download files from remote storage
- `common/lora_utils.py` --> should include generic abstractions that make it easy to download lora adaptors. It uses abstractions from download_utils.py

- `llm/serve` --> Different parts of llm serve that need lora or generic model download use above abstractions

- `llm/batch` --> similarly different parts of llm/batch that need to use lora or remote download use the above abstractions. 

It should be super clean to maintain and extend after this PR. You should not see weird leakage of dependencies and concepts between batch and serve or between common <> serve / batch

nrghosh committed:
WIP: address feedback, re-refactor, lint

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
WIP - fix imports

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
WIP - full top-down refactor

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
WIP - fix imports post refactor

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
Merge branch 'master' into refactor-lora_model_loader

Signed-off-by: Nikhil G <nrghosh@users.noreply.github.com>
Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh commented:
Re: points in thread ^

> **1. "level of refactoring"**
** Comprehensive top-down refactor **

- **`common/download_utils.py`** - Generic file downloading abstractions (no LoRA-specific concepts)
- **`common/lora_utils.py`** - Generic LoRA abstractions using download utilities  
- **`common/models.py`** - Shared data structures (`DiskMultiplexConfig`, `global_id_manager`)
- **Serve/Batch** - Use common abstractions without circular dependencies

** Result**: Clean separation with no leakage between batch/serve or common/specific concerns.

>**2. "go generic → specific"**
**Restructured as suggested:**

1. **Generic primitives first**: `download_utils.py` provides pure file downloading
2. **LoRA abstractions**: `lora_utils.py` uses generic download utilities
3. **Specific implementations**: Serve/batch use the generic abstractions
4. **No weird leakage**: Clean dependency flow with no circular imports

> **3. "clean to maintain and extend after this PR"**
- **No circular dependencies** - Clean import hierarchy
- **Generic → specific architecture** - Easy to extend for new use cases
- **Centralized utilities** - Single source of truth for download logic
- **Backward compatibility** - Existing code continues to work

> **"Serve test failures are inclusive due to..."**
**Confirmed**: The test failures we're seeing are **not** due to our refactoring:
- Ray Serve configuration issues (`env_vars` not allowed in `ray_actor_options`)
- These would fail regardless of our LoRA refactoring
- Our core functionality tests pass 

**Refactor**
1. Moved `LoraModelLoader` to `download_utils.py` as `_LoraModelLoader` (private)
2. Enhanced `download_lora_adapter` with unified functionality
3. Centralized download utilities in `common/utils/`
4. Unified retry logic, caching, and file locking
5. de-duplication while maintaining backward compatibility
6. Fixed all circular imports and dependency issues

## **Tests Passing**:
- LoRA Model Loader Tests 
- Import functionality 
- Core refactoring functionality 

### **Architecture Benefits**:
- Generic → specific as requested
- No dependency leakage between batch/serve
- Easy to maintain and extend
- Clean separation of concerns

nrghosh committed:
fix lint

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
fix: implement parse_args and fix env_vars placement

- Implement LLMServingArgs.parse_args() to properly parse YAML files to LLMConfig objects
- Add _parse_path_args() and parse_args() functions for string->LLMConfig conversion
- Move env_vars from ray_actor_options to runtime_env to fix Ray Serve validation
- Resolves AttributeError: 'str' object has no attribute 'model_id'
- Resolves ValueError: Specifying 'env_vars' in ray_actor_options is not allowed

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
fix: suppress Pydantic model_id field namespace warning

- Add model_config with protected_namespaces=() to DiskMultiplexConfig
- Resolves: Field 'model_id' in DiskMultiplexConfig has conflict with protected namespace 'model_'

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
wip refactor + CI test fixes - consolidate lora utilities, remove dupe functions, update serve mux utils to use canonical funcs

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
fix lora model listing issue and import concistency

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
Merge branch 'master' into refactor-lora_model_loader

Signed-off-by: Nikhil G <nrghosh@users.noreply.github.com>

nrghosh committed:
wip - lint

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
final refactor - duplicated LoRA utilities

- consolidate some utiliites into canonical location
- update imports

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
update imports

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

kouroshHakha commented on python/ray/llm/_internal/common/utils/download_utils.py:
```diff
@@ -316,7 +333,195 @@ def download_lora_adapter(
     return downloader.get_model(tokenizer_only=False)
+
+
+class _LoraModelLoader:
```
`download_lora_adapter` and `_LoRAModelLoader` need to be moved to `lora_utils.py`?

kouroshHakha reviewed (approved):
Everything checks out except the location. I'll fix and merge.

kouroshHakha committed:
wip

Signed-off-by: Kourosh Hakhamaneshi <kourosh@anyscale.com>

kouroshHakha reviewed (approved):
LGTM. waiting for release tests: https://buildkite.com/ray-project/release/builds/48074
to make sure everything passes.

kouroshHakha commented on python/ray/llm/_internal/common/utils/lora_utils.py:
```diff
@@ -0,0 +1,486 @@
+            lora_mirror_config.bucket_uri,
+            model_local_path,
+            timeout=self.download_timeout_s,
+            sync_args=lora_mirror_config.sync_args,
```
It's true, it can be remove I think?

kouroshHakha commented on python/ray/llm/_internal/common/utils/lora_utils.py:
```diff
@@ -0,0 +1,486 @@
+        # Apply retry decorator to _download_lora at runtime with instance parameters
+        download_with_retries = retry_with_exponential_backoff(
+            max_tries=self.max_tries,
+            exception_to_check=Exception,  # Catch any exception from CloudFileSystem
```
keep Exception, it's fine.

nrghosh committed:
fix ci (patch target -> lora_utils)

- update CloudFileSystem patch target properly to mock
  CloudFileSystem.download_files() and prevent S3 calls in testing
- caused by moving _LoraModelLoader from download_utils to lora_utils.
- tests were patching CFS in download_utils instead lora_utils

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

kouroshHakha commented:
@nrghosh the peremerge failed unfortunately.

nrghosh commented on python/ray/llm/_internal/common/utils/lora_utils.py:
```diff
@@ -0,0 +1,486 @@
+def clear_directory(dir: str) -> None:
+    """Clear a directory recursively, ignoring missing directories."""
+    try:
+        subprocess.run(f"rm -r {dir}", shell=True, check=False)
```
ack

nrghosh commented:
@kouroshHakha ci tests (lora test functional ones in llm-cpu specifically) fixed, but new issue appeared with ubuntu version / loading docker plugin

- [wanda: oss-ci-base_cu128-py3.11
](https://buildkite.com/ray-project/premerge/builds/43415/steps/table?jid=0197f13e-ee40-4d26-b1eb-68144bdd574f)
- [discussion](https://anyscaleteam.slack.com/archives/C02J07KKF2A/p1752097593471359)

nrghosh committed:
remove unused sync_args param

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
Merge branch 'master' into refactor-lora_model_loader

kouroshHakha commented:
@nrghosh lora release test is not passing. Can you take a look? TY

nrghosh committed:
Fix CI -  race condition in LoRA router init

- Refactoring that moved LoRA util functions from multiplex/utils.py to
  lora_utils.py changed some of the initialization timing

- Exposed an existing race condition in the router async init process

- CI test llm_serve_llama_3dot1_8B_lora release test was failing with
  all tests returning `openai.NotFoundError: {"error": Requested
resource not found"}`

- Target model was returning 404 despite being properly configured

- Router __init__ method populates _llm_configs async, but three methods
  access _llm_configs without waiting for initialization to complete

- Fix: added await self._init_completed.wait() to all three of the
  methods (model(), models(), and _get_response())

- This way, they wait for async initialization to complete before
  accessing _llm_configs

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
Merge branch 'master' into refactor-lora_model_loader

Signed-off-by: Nikhil G <nrghosh@users.noreply.github.com>

kouroshHakha commented on python/ray/llm/_internal/serve/deployments/routers/router.py:
```diff
@@ -282,6 +284,7 @@ async def _get_response(
         Union[LLMChatResponse, LLMCompletionsResponse, LLMEmbeddingsResponse], None
     ]:
         """Calls the model deployment and returns the stream."""
+        await self._init_completed.wait()
```
this doesn't make sense, why is this change needed? The main changes in this PR should only affect the LLMServer not the LLMRouter layer

nrghosh replied:
the refactor (even though it's unrelated to LLMRouter directly) caused a change in module timing - which resulted in unexpected behavior from LLMRouter layer. 

__init__ for Router returns immediately but kicks off an async initialization loop in the background - before the refactor, `get_lora_model_ids` was imported as part of the serve deployment module loading. But now that it's in `lora_utils.py`, that's a new import dependency. So the timing for the import+call is changed which means first request can arrive while configs are still empty .

nrghosh committed:
lint

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
revert race condition hypothesis

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
fix (release) - consolidate lora download paths

- refactor caused inconsistent download mechanisms which made serve
deployment integration tests fail.

- both batch and serve deployments now use the same download functionality

- modified _LoraModelLoader to use canonical download_lora_adapter()

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
fix: resolve LoRA download validation errors and restore working implementation

- Revert _download_lora to use CloudFileSystem.download_files directly
- Fix test fixture bucket_uri to include lora_id path component
- Restore llm_server.py to use proper _LoraModelLoader async interface
- Eliminates CloudMirrorConfig validation errors caused by invalid S3 URIs
- All unit tests now pass (8/8 vs previous 2/8)
- Should resolve release test hanging issues

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
fix: restore LoRA configuration parsing logic that was broken during refactor

- Restore get_lora_finetuned_context_length() to properly parse adapter_config.json
- Fix download_multiplex_config_info() to actually call config parsing instead of returning defaults
- Remove duplicate function from multiplex/utils.py to eliminate circular imports
- Fixes the core integration issue causing release test failures/hangs
- Release test (llm_serve_llama_3dot1_8B_lora) failing  wasn't caught
  because unit tests bypass config parsing logic

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
Add unit tests for LoRA config parsing logic

Add TestLoRAConfigurationParsing class with tests for:
- get_lora_finetuned_context_length() with valid JSON parsing
- Missing config file handling
- Invalid JSON error handling
- download_multiplex_config_info() integration
- Default fallback behavior

This addresses the coverage  gap that allowed the configuration
parsing regression to reach release tests undetected. The original
tests only mocked CloudFileSystem.download_files but never tested
the CloudFileSystem.get_file path used for config parsing.

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>
